### PR TITLE
PERF: avoid is_bool_indexer check where possible

### PR DESCRIPTION
**Pull request by jbrockmendel**
xref #30349, not a lot of ground to pick up here, but we can avoid a few calls

### CODE REVIEW

File: pandas/core/series.py [hunk 1]
```diff
@@ -856,31 +856,36 @@ def __getitem__(self, key):
...
+            try:
+                result = self.index.get_value(self, key)
+
+                return result
+            except InvalidIndexError:
                 pass
-            else:
+            except (KeyError, ValueError):
+                if isinstance(key, tuple) and isinstance(self.index, MultiIndex):
+                    # kludge
```
Comment (jorisvandenbossche) [inline pandas/core/series.py]: Such a comment is not really helpful if you don't explain the kludge ..

**Review by jorisvandenbossche — COMMENTED** [review]
Can you show some timings?

File: pandas/core/series.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/series.py]: agreed, but this comment is present in the status quo

File: pandas/core/series.py [hunk 2]
```diff
@@ -856,31 +856,36 @@ def __getitem__(self, key):
...
-            pass
-        except (KeyError, ValueError):
-            if isinstance(key, tuple) and isinstance(self.index, MultiIndex):
-                # kludge
-                pass
-            elif com.is_bool_indexer(key):
+        if key_is_scalar or isinstance(self.index, MultiIndex):
+            # Otherwise index.get_value will raise InvalidIndexError
+            try:
+                result = self.index.get_value(self, key)
```
Comment (jreback) [inline pandas/core/series.py]: prefer you just do a return rather than assign and return

**Review by jreback — COMMENTED** [review]
does this make any difference or just moving code around?

**Comment by jbrockmendel** [comment]
> does this make any difference or just moving code around?

This should not change any behavior.

> prefer you just do a return rather than assign and return

I like doing it in two lines inside try/excepts so i can see in coverage measurements whether the call is ever successful

**Comment by jreback** [comment]
can you rebase; i am not sure if @jorisvandenbossche had more comments.

**Comment by jorisvandenbossche** [comment]
Just my question from above: Can you show some timings?

**Comment by jbrockmendel** [comment]
> Can you show some timings?

It's pretty small (especially since is_scalar and is_iterator were recently optimized)

```
In [1]: import pandas as pd 
   ...: import numpy as np 
   ...:  
   ...: ser = pd.Series(range(10**5)) 
   ...:  
   ...: indexer = np.random.random(10**5).astype(bool) 
   ...:                                                                                                                                                                                                                                           

In [2]: %timeit ser[indexer]                                                                                                                                                                                                                      
220 µs ± 3.93 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)  # <-- master
213 µs ± 9.02 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)  # <-- PR
```

**Comment by jorisvandenbossche** [comment]
And with a scalar indexer?

**Comment by jbrockmendel** [comment]
for the scalar case I expect the difference to be zero, or even slower by the amount of a `isinstance(self.index, MultiIndex)` check

```
In [4]: %timeit ser[999]                                                        
1.85 µs ± 81.1 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)  # <-- master
1.82 µs ± 57.6 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)  # <-- PR
```

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
> or even slower 

That's what I expected as well.

(in general, if you do changes specifically for performance, it's helpful for reviewing that you show some relevant timings)

**Comment by jreback** [comment]
@jorisvandenbossche pls make sure labels and versions on the issue and PR are good before merging

**Pull request merged**